### ISSUE TITLE
Master losing connection to signaling channel (using C SDK 1.3)

### ISSUE THREAD

**Issue by modescode** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#729]
A customer is running into failure cases, when they try to stream more than 2 times. The first two times viewer connects and streaming works, where as on subsequent attempts the live streaming fails because the connection from master to signaling channel has been lost. Once the device goes into this state, it does not recover unless we reboot it. 

This symptom only started occurring after the ODM integrated with SDK version 1.3. Logs attached.

[device_logs_master_losting_connection.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5110521/device_logs_master_losting_connection.log)

**Comment by hsinghaowang** [comment]
sometimes the **signalingClientState** changed to **SIGNALING_CLIENT_STATE_GET_ICE_CONFIG**
and it seems  not doing anything to get back to **SIGNALING_CLIENT_STATE_CONNECTED**.
What can I do if this happen? except restart ?

**Comment by fernandoc1** [comment]
This is also what happens to me.
[output.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5117516/output.log)

**Comment by MushMal** [comment]
@fernandoc1 The refreshing of the ICE candidate happens in the background and the state machine is getting into connected state. There is, however, an error sending a message: STATUS_SIGNALING_DUPLICATE_MESSAGE_BEING_SENT

I am not sure what could cause this or what modifications you might have made (the log snippet doesn't seem to capture it) but there is a chance to receive this message if the sending of the previous message is in progress while a new message is being send. 

@hsinghaowang could you please capture the logs so we can see what you are seeing? The stock master sample application will attempt to re-create the signaling client on error - those are normal conditions when for example the connection gets interrupted and we are unable to reconnect within the given timeout (up-to 5 attempts within 15 seconds).

@modescode have you had a chance to get more information as I requested?

**Comment by fernandoc1** [comment]
@MushMal, I have no problem of showing my code changes.
I'm just compiling the version 1.3.0 and I have just used the master example, with changes in the function that reads frames from disk. I just make that function read frames from my shared memory source. The rest of the code is exactly the same.

**Comment by MushMal** [comment]
@fernandoc1 I would recommend opening a separate issue to track the ICE connectivity issue on your network topology. There are some tooling that we have built-in that could help troubleshooting your e2e connection between the master and viewer and understand whether it's attempting to use P2P or TURN, etc..

However, this issue is about the signaling connection

**Comment by modescode** [comment]
@MushMal  - I'll try to get answers from the customer today.

**Comment by hsinghaowang** [comment]
@modescode I am your customer.

**Comment by MushMal** [comment]
LOL :) @hsinghaowang could you please respond to the items I have requested?

**Comment by modescode** [comment]
Ha! :) @hsinghaowang thanks for the call out. Please answer @MushMal directly.

**Comment by MushMal** [comment]
BTW, GitHub issues is the surest and the fastest way to get help that can benefit other members of the community too.

**Comment by hsinghaowang** [comment]
Here is the log (total 5 minutes logs):
[webrtc_1.3_state6.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5121537/webrtc_1.3_state6.log)

in line 2640, it changed state 6 (SIGNALING_CLIENT_STATE_GET_ICE_CONFIG), and never come back to 9 (SIGNALING_CLIENT_STATE_CONNECTED) after about 2 hours.
[08-25 12:00:29] [11765][W]webrtc_master: heartbeat. state:[6], sesssions:[1]

I print the state in main loop sessionCleanupWait()
```
...
        // Check the signaling client state and connect if needed
        if (IS_VALID_SIGNALING_CLIENT_HANDLE(pSampleConfiguration->signalingClientHandle))
        {
            SIGNALING_CLIENT_STATE signalingClientState;
            MY_CHK_STATUS(signalingClientGetCurrentState(
                pSampleConfiguration->signalingClientHandle,
                &signalingClientState)
            );
            if (signalingClientState == SIGNALING_CLIENT_STATE_READY)
            {
                UNUSED_PARAM(signalingClientConnectSync(pSampleConfiguration->signalingClientHandle));
            }
            LOG_W("webrtc_master: heartbeat. state:[%d], sesssions:[%u]",
                  signalingClientState,
                  pSampleConfiguration->streamingSessionCount
            );
        }
        else
        {
            LOG_E("webrtc_master: heartbeat. Invalid signaling client handle");
        }
...
```

**Comment by hsinghaowang** [comment]
Here is another log which run into state 8 and not come back. see line 3966, it changed to state 8.
I just run 1 session to test the memory usage, then it just happened.

[webrtc_1.3_state8.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5123041/webrtc_1.3_state8.log)

**Comment by hsinghaowang** [comment]
If we want to reduce memory usage, any suggestion?

**Comment by MushMal** [comment]
I have been looking at the first log and I can't really pinpoint yet what's going on. In LN 2580, the signaling client fails to connect with 

[08-25 12:00:25] connectSignalingChannelLws(): operation returned status code: 0x0000000f

This happens after a number of attempts to reconnect. The signaling client then issues an error on which the stock master application re-creates the signaling client.

The other question is why did the signaling client attempt to reconnect. The signaling client attempts to keep the connection open at all times to minimize the latency - this is the reason why it has a handler for reconnect. If the connection is interrupted, it will attempt to reconnect having a 5 second default timeout per connection attempt with overall 15 seconds timeout. If this is hit, it will stop retrying and will issue an error callback.

In this example, the connection got dropped either by the time limit enforced by the signaling backend or a networking error (there are no debug logs)

[08-25 12:00:11] onConnectionStateChange(): New connection state 4

The signaling client then attempts to reconnect and moves to Describe state and iterates:

08-25 12:00:13] stepStateMachine(): State Machine - Current state: 0x0000000000000100, Next state: 0x0000000000000002
[08-25 12:00:13] stepStateMachine(): State Machine - Current state: 0x0000000000000002, Next state: 0x0000000000000004
[08-25 12:00:13] stepStateMachine(): State Machine - Current state: 0x0000000000000004, Next state: 0x0000000000000010
[08-25 12:00:13] stepStateMachine(): State Machine - Current state: 0x0000000000000010, Next state: 0x0000000000000020


The Describe and getEndpoint states are cached so they evaluate promptly

[08-25 12:00:20] stepStateMachine(): State Machine - Current state: 0x0000000000000020, Next state: 0x0000000000000040
[08-25 12:00:20] stepStateMachine(): State Machine - Current state: 0x0000000000000040, Next state: 0x0000000000000080

This is when the state machine executes Connect state but it fails with the timeout.

[08-25 12:00:25] connectSignalingChannelLws(): operation returned status code: 0x0000000f

(NOTE: The single attempt which didn't resolve within the 5 seconds).
It iterates over again but for some reason the application terminates or starts a teardown. This is the end of the log so I am not sure what's going on.

**Comment by MushMal** [comment]
Looking at the second log around LN 3960 as you've mentioned, there is a similar story, just around this time the connection got dropped and the signaling state re-connects OK:

[08-25 17:52:15] lwsWssCallbackRoutine(): Connection established

Are you listening to state change callback? Can you print it out in your log? The logs are saturated by frame drop and rtp packet received logs which are super verbose and make it hard to distill the scenario out of it. We have been working on making the rtp output less verbose.

**Comment by hsinghaowang** [comment]
@MushMal Yes, I am listening to state change callback and print it.
You can search "New connection state".

the code is:
```
VOID
onConnectionStateChange(
    UINT64 customData,
    RTC_PEER_CONNECTION_STATE newState)
{
    auto pSampleStreamingSession = reinterpret_cast<PSampleStreamingSession>(customData);
    MY_ASSERT_OR_RETURN(pSampleStreamingSession != nullptr);
    
    DLOGI("New connection state [%u]", newState);

    if (newState == RTC_PEER_CONNECTION_STATE_FAILED ||
        newState == RTC_PEER_CONNECTION_STATE_CLOSED ||
        newState == RTC_PEER_CONNECTION_STATE_DISCONNECTED)
    {
        ATOMIC_STORE_BOOL(&pSampleStreamingSession->terminateFlag, TRUE);
//        CVAR_BROADCAST(pSampleStreamingSession->pSampleConfiguration->cvar);
        pSampleStreamingSession->pSampleConfiguration->cvar.notify_all();
    }
}
```

**Comment by MushMal** [comment]
That's RTC Peer connection state changes. You need to be listening to the signaling client state changes:
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/include/com/amazonaws/kinesis/video/webrtcclient/Include.h#L1235

**Comment by jonescx** [comment]
@MushMal I use webrtc sdk 1.3.0 and modify ice policy to relay only at 
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/samples/Common.c#L372

`configuration.iceTransportPolicy = ICE_TRANSPORT_POLICY_RELAY;`

I use [KVS WebRTC Test Page](https://awslabs.github.io/amazon-kinesis-video-streams-webrtc-sdk-js/examples/index.html) to verify the streaming and it works in the beginning after a few hours. It fails to view the streaming. 
Here are two logs for success and fail cases. Could you help to check the attached log? Thanks.

[webrtc_1.3_relay_success.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5128139/webrtc_1.3_relay_success.log)
[webrtc_1.3_relay_fail.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5128140/webrtc_1.3_relay_fail.log)

**Comment by hsinghaowang** [comment]
@MushMal Here is another log with "Signaling client state changed"
at line 1408 it changed to 8 and never come to 9.

[webrtc13_state8_08_26.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5128193/webrtc13_state8_08_26.log)

**Comment by sbarry157** [comment]
After discussion with @jonescx and @modescode, this ticket is chasing two different issues. I created a separate ticket to track the issue of the camera losing connection to the signaling channel with no memory issues here #752

**Comment by MushMal** [comment]
@hsinghaowang, I was looking at webrtc13_state8_08_26.log and in LN 1500

`[08-26 14:57:57] connectSignalingChannelLws(): operation returned status code: 0x0000000f`

So the terminated connection failed to reconnect. What I am puzzled about is that I don't see any printouts from:

https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/samples/Common.c#L110

This should have been called as the error handler from: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Signaling/LwsApiCalls.c#L1318


Folks, this is the 3rd back and forth related to this issue. How can I help you to debug these things better? It's not very productive to dig through tons of application logs to understand the scenarios. In this particular case, have you modified the error handler routine at all? What are the debug steps you are taking?

From the logs I see a ton of issues with your networking connection - likely due to the PHY layer of the wifi.

**Comment by jonescx** [comment]
@MushMal 

> This should have been called as the error handler from: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Signaling/LwsApiCalls.c#L1318
`                pSignalingClient->signalingClientCallbacks.errorReportFn(pSignalingClient->signalingClientCallbacks.customData,
                                                                         STATUS_SIGNALING_RECONNECT_FAILED, reconnectErrMsg, reconnectErrLen);`

The errorReportFn is assigned as signalingClientError() as shown below, it'd print logs.
```
STATUS
signalingClientError(
    UINT64 customData,
    STATUS status, PCHAR msg, UINT32 msgLen)
{
    auto pSampleConfiguration = reinterpret_cast<PSampleConfiguration>(customData);

    DLOGW("Signaling client generated an error [0x%08X] - '%.*s'", status, msgLen, msg);

    // We will force re-create the signaling client on the following errors
    if (status == STATUS_SIGNALING_ICE_CONFIG_REFRESH_FAILED
        || status == STATUS_SIGNALING_RECONNECT_FAILED)
    {
        ATOMIC_STORE_BOOL(&pSampleConfiguration->recreateSignalingClient, TRUE);
//        CVAR_BROADCAST(gSampleConfiguration->cvar);
        gSampleConfiguration->cvar.notify_all();
    }

    return STATUS_SUCCESS;
}
```

Recently, we patched two fixes and build debug version of libwebsocket try to track memory usage issue.
* fix race condition in sample on sampleStreamingSessionList #a1a5a451a3c24f33b3165e87f303a021cb438324
* Fix bugs in sample  #f27375f9ee647c713dcb66f05664059254c0cf6f

**Comment by MushMal** [comment]
@jonescx it's not clear why your error function printout doesn't output anything on the error in the logs or whether it's been called. 

The timeout happens on reconnect which is shown in the logs. 

To help debug your case, could you do the following:

* How often does this happen? Clearly, your network has an outage causing to reconnect but how often the reconnect fails?
* Can you run the latest master branch master sample on your device with no modifications? 
* Could you add some instrumentation in the form of a printout around https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Signaling/LwsApiCalls.c#L1215 terminate API call. Make sure you output different text so we can trace what has been called. You could also run it under the debugger if you have one and set breakpoints. NOTE: The only blocking is gated by a timeout of 5 seconds here.
* Add instrumentation around https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Signaling/LwsApiCalls.c#L1318 to see whether the callback is called.

I am not sure how effective we can be debugging these issues unless we can reproduce any of these locally which we are unable to do.

**Comment by jonescx** [comment]
@MushMal Thanks for your advise.

We are trying do the 2 things
1. Setup a minimum modification version 
2. Find out where makes the webrtc state stuck.
The state stuck issue is not easy to reproduce, it may take long time to streaming.

under editing..but put the log for discussion first.
[0901_webrtc_stuck_at_state6.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5153505/0901_webrtc_stuck_at_state6.log)

**Comment by MushMal** [comment]
Sounds good. Meanwhile, could you parallelize the modified version with the stock version of the master and ensure that works ok? I just don't know how else to help as without a repro we can't help in a more meaningful way. The log you've attached has Info level only enabled and for some reason I see no printout from the callback but the state machinery proceeds further. It makes no sense as the calls to the state transition callback are synchronous.

**Comment by jonescx** [comment]
Thanks for giving suggestions today.
We turn off the async ICE server config 
`pSampleConfiguration->channelInfo.asyncIceServerConfig = FALSE;`

And also try to lock the location that makes the thread stuck.
About the jitter buffer memory save commits, we will apply it after we confirm the commit number.
I guess it's the following one but I am not sure. 
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/commit/a06e9eaf2c2a251e34e6687c21fa4ff3a07f0285

**Comment by MushMal** [comment]
Looks like it. You could also take the latest

**Comment by jonescx** [comment]
I reproduced the thread stuck issue again. 

[0902_webrtc_stuck_state6.log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5159652/0902_webrtc_stuck_state6.log)

**Comment by MushMal** [comment]
@jonescx this is an interesting log. Have been looking at this for quite some time now. Do you have the full log that can give us a little more details? In this case, it seems that there was a disconnect detected prior (not captured in the log) and the reconnect handler is attempting to connect. In this case, it fetches the credentials which hits a cache and then proceeds with other states which are also cached. The first non-cached state then is refreshing ICE config. This results in a number of failures due to networking (see the errors with SSL). However, in LN 547 I see the entry (garbled as it is intermixed with other output) freeLws which is likely the output from getIceConfigLws() API teardown calling freeLwsCallInfo(). This will then terminate the state machine step call in the reconnectHandler() routine which is executed in a new thread that was created to handle the reconnect on error.

What's really bizarre is that it should have called the error handler in LwsApiCalls.c LN 1318. 

https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Signaling/LwsApiCalls.c#L1308-L1330

I am not sure what's going on but as I recommended, could you add a lot of instrumented output in this CleanUp section in the reconnectHandler API to understand what the code is doing when you hit this case? I want to make sure that we are actually hitting the case connect and if not then why we are not calling the error handler which would then iterate and re-create the state machinery.

What I am not sure I understand is that if the theory is correct, then the reconnectHandler exits and if it exits then there has to be an error (retStatus should be non STATUS_SUCCESS) in which case the error callback should be called. The error callback that's provided in Common.c should output into the log.

**Comment by jonescx** [comment]
@MushMal I upload the full log from the webrtc starts. I open the test page viewer and let it stays overnight. The viewer is working fine when I check it in the morning. However, the signaling thread is blocked as you see in the short one. You can check the "0902_extra_log_patch.txt" to know where the extra logs comes from. My code base is the tag 1.3.0.

I will add more instrumented output to figure out what happen in the reconnectHandler().
Thanks for the advise.

[0902_extra_log_patch.txt](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5160779/0902_extra_log_patch.txt)
[0902_webrtc_stuck_state6_full.zip](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5160759/0902_webrtc_stuck_state6_full.zip)

**Comment by jonescx** [comment]
@MushMal I check the system log since the output won't be scrambled in the log file. I found the errorReportFn is NULL in the reconnectHandler(). Please check the zip file for log_server2 at line 14898. 

[0902_logs_server.zip](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5161009/0902_logs_server.zip)

**Comment by MushMal** [comment]
@jonescx if the errorReportFn is set to NULL then everything can be explained! This simply means that the signaling has failed to reconnect due to networking and gave up. The application supplying the error handler can react to it. The default master sample does exactly that. This is the reason I asked whether you are running stock application.

**Comment by jonescx** [comment]
@MushMal I see. We will check why the errorReportFn is NULL. The attached is the new log with more outputs for freeLwsCallInfo() and reconnectHandler() in LwsApiCalls.c. The two functions leave normally. So it means the thread is not blocked and it just give up, right?
[0903_webrtc_stuck_state6_short.txt](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5165645/0903_webrtc_stuck_state6_short.txt)

**Comment by MushMal** [comment]
Had looked at the latest log 0903_webrtc_stuck_state6_short.txt

The log starts with a timeout and the lws callback handler kicks off the reconnect routine as your signaling client is configured with reconnect. This is good. The reconnect routine then primes the state machine - as you can see the state machine gets driven by the reconnect thread (which was just created) and it moves past the few cached states and ends up with the getIceConfig. Here, it attempts to reach out to the backend but due to the networking issue it times out - this takes approximately 20 seconds which blows past the timeout which is 15 seconds. The reconnect handler then goes through the terminate sequence and I see your instrumented output there. It indeed has a non-successful retStatus and the structures are not NULL so it attempts to call the error callback to indicate to the application that it is giving up after the retries. As the error handler function is null the caller application doesn't handle it like the sample master application does. In the sample, we set an appropriate flag and retry later in the outer loop in Common.c.

In your case, everything actually works just fine - it's a negative case with a network error (perhaps a transient one) which then your application doesn't handle as you don't specify an error handler.

It's trivial to reproduce in your case - just run your application on the device and terminate the networking. You will see this behavior which is by design - the signaling client doesn't keep retrying infinitely. It's the master sample that retries upon the signaling client reconnect retry exhaustion.

Please give this a try to reproduce first - you should be able to reproduce this within 10 seconds (after the pong failure). Please run the stock master application which has the recovery logic based on the error handler - this should at least keep retrying. Add the recovery to your code and this should solve the issue.

**Comment by jonescx** [comment]
@MushMal Thanks your reply. I can easily reproduce it by stopping networking. We also found the location to set the errorReportFn as NULL. We will fix it and try again.

**Comment by MushMal** [comment]
Perfect! Looking forward for the results

**Comment by modescode** [comment]
@MushMal - I'm running into an intermittent timeout on "SendAlexaOfferToMaster" (6s). The master logs show there is no offer received on the master for those requests. See master logs (redacted channel ARN) - 
[log_server3 (1).log](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5171180/log_server3.1.log)

I pinged you the channel name separately.

**Comment by MushMal** [comment]
@modescode perhaps those are related but until the issue with the error handler is not fixed it makes no sense in investigating any time as any network issue lasting longer than 15 seconds will cause the signaling client to stop retrying to reconnect

**Comment by modescode** [comment]
Hi @MushMal  - we received latest build this morning, which fixes the signaling issue and the above testing was done on that build. @jonescx - can you confirm that the 6.9E build shared this morning fixes the signaling issue that was discussed 19 hours ago?

**Comment by MushMal** [comment]
@modescode I looked at log_server3 (1).log and this time I don't see any issues with the signaling - we have a few transient network faults that the SDK is able to recover from. It's always connected. There are two (I believe) sessions in the log where the offer has been answered and some frames have been exchanged. I see a connection termination in LN 6709

`[ 725][D]socketSendDataWithRetry(): sendto() failed with errno Connection reset by peer`

after which we get errors sending the packet. This might be an issue if the application is not handling the peer connection state change notification properly. 

At any rate, this is an orthogonal issue and needs to be looked at separately.

@chehefen do you have any ideas?

**Comment by jonescx** [comment]
@modescode @MushMal I tested the new firmware with errorReportFn fixed version. The signal channel keeps stay at [9] state and live streaming is working as well. I don't see signal state the stuck at [6] issue. I think this issue is fixed. Thank you.

**Comment by disa6302** [comment]
@jonescx , 

Closing this issue based on your comment. Do not hesitate to reach out with any questions in the future.
